rfcs/inline_symbols.md

```diff
@@ -0,0 +1,521 @@
+can be used as a symbol, field name, or annotation, it *must* be added to the symbol table first.
+
+This constraint incurs a number of expenses:
+
+* Writers must examine each top-level value that they intend to emit (however deeply nested) to
+  determine which of its symbols, field names, and annotations are not yet in the symbol table.
+* If any such elements are found, a new Local Symbol Table (LST) or LST-append must be written
+  to the stream establishing a new symbol ID for each of those elements.
+* Writing the top-level value itself must be deferred until the new symbols are added to the table, 
+  necessitating additional buffering of encoded output.
+* Both Readers and Writers must hold the complete symbol table in memory. The more entries there
+  are in the symbol table, the more memory it will consume.
+* Symbol IDs are encoded as variable-length unsigned integers. The more symbol IDs there are, the
+  more bytes it will take to encode the newest symbol IDs.
+  
+The symbol ID encoding requirement is designed to amortize the cost of writing out the complete text
+of recurring text elements over the lifetime of the Ion stream; by paying an up-front cost to
+detect and define new symbols, all future references to the same text element can be made
+substantially cheaper to write. These overhead costs are worthwhile for symbols that are frequently
+referenced. However, there are a variety of use cases for which the costs are never recouped.
+
+### Short-lived streams
+
+Consider this short-lived Ion text stream representing data collected from a weather station:
+
+```js
+{
+  sensorId: 12345,
+  type: sensorData,
+  reading: {
+    temperature: celsius::12.5,
+    time: 2020-10-22T16:00:00Z
+  }
+}
+```
+
+When the same data is re-encoded in binary, all of the symbols, struct field names, and annotations
+must be added to a local symbol table definition that appears at the beginning of the stream.
+
+```js
+$ion_1_0
+$ion_symbol_table::{ // Add all of our symbols, field names, and annotations to the symbol table
+  sensorId,   // $10
+  type,       // $11
+  sensorData, // $12
+  reading,    // $13
+  temperature,// $14
+  celsius,    // $15
+  time,       // $16
+}
+{
+  $10: 12345, // Then encode our symbols, field names, and annotations as symbol IDs
+  $11: $12,
+  $13: {
+    $14: $15::12.5,
+	$16: 2020-10-22T16:00:00Z
+  }
+}
+```
+
+The symbol ID encoding requirement forces us to process our sensor data twice: once to detect any
+symbols, annotations, and field names that needed to be added to the symbol table and once to write
+the sensor data itself. However, because none of the values being added to the symbol table appear
+more than once, we do not benefit from the compactness of the symbol ID encodings. The overhead
+associated with adding each of them to the symbol table slows down the writing process and inflates
+the overall size of the data.
```

Comment:
> What about this example makes it so that shared symbol tables is not a better solution?

When you say "this example" are you referring to the example below, in _Short-lived streams_?

Shared Symbol Tables (SSTs) address a small portion of the problem. Specifically, they make it possible to omit LST appends for _expected symbols,_ which can save on data size. (All of the symbols used in _Short-lived streams_ are expected symbols, so an SST would work well there.)

However, writers have no way of knowing whether the active symbol table (shared, local, or system) contains _all_ of the symbols, field names, and annotations that they will be asked to write. As such, the full 'discovery' phase of writing still has to happen for every value, even if it happens to come up empty-handed each time. For many Ion streams, it's impossible to define all of the symbols you'll need in advance. Serializing maps is likely the biggest driver of this, as the set of struct fields that might be referenced is unknown until runtime.

Additionally, using SSTs involves a commitment to some amount of additional `Catalog` infrastructure. Ion streams relying on SSTs are no longer self-contained, so sharing Ion datagrams becomes more complex. Readers on other hosts must have a way of locating and accessing the declared SST dependencies to read the data. This extra coordination can be worthwhile, but it creates both friction and pitfalls for simpler use cases.

I'll add a paragraph about this at the end of _Short-lived streams_ illustrating the distinction between expected and unexpected symbols.

Reply:
I've [added a new section](https://github.com/amzn/ion-docs/pull/137/commits/5230db4896768132f506ba7bbd9ba3d96e51982e#diff-748521366873d3b7ee89e9d8ef0b1dbff910dd869c817aa9296284a826e2ac0b) to attempt to address this.